### DISCUSSION TITLE
How to support a row-type function parameter with an indeterminate number of columns？

### DISCUSSION THREAD

**Discussion by duanmeng** [facebookincubator/velox#6403] [Q&A]
I am trying to add a "to_json" function in pr #6287, which supports 3 input types, `ARRAY`, `MAP`, `ROW`. It seems to only support a fixed number of columns, such as `.argumentType("row(bigint, varchar)")`. But we **cannot** determine how many columns there are especially the type of each column at definition time. I cannot find a way to support this scenario although I can use `.argumentType("T")`.

**Comment by mbasmanova** [comment]
CC: @kagamiori 

This is not supported yet. Curious, why to_json function supports only ARRAY, MAP and ROW. I'd expect it to support all types.

**Reply by duanmeng** [reply]
Spark only support array, map and struct (aka row in velox), it [checks the input types](https://github.com/apache/spark/blob/a0a78b6081f165c78de3b776cefc01c5e5c9a680/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala#L733), and Velox's [JsonCastOperator checks and only support array, map and row](https://github.com/facebookincubator/velox/blob/c79be9bcdf19cf5ecb7b581429fc9b2fb541dfb6/velox/functions/prestosql/types/JsonType.cpp#L788) as well.

SparkSQL
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala#L733
```Java
  override def checkInputDataTypes(): TypeCheckResult = inputSchema match {
    case struct: StructType =>
      ...
    case map: MapType =>
      ...
    case array: ArrayType =>
      ...
    case _ => TypeCheckResult.TypeCheckFailure
```

Velox
https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/types/JsonType.cpp#L788
```C++
bool JsonCastOperator::isSupportedToType(const TypePtr& other) const {
  ...
  switch (other->kind()) {
    case TypeKind::ARRAY:
      return isSupportedToType(other->childAt(0));
    case TypeKind::ROW:
      ...
      return true;
    case TypeKind::MAP:
     ...
    default:
      return false;
```
